### PR TITLE
DEVPROD-31985 Revert test result upload partition to when results are processed

### PR DESCRIPTION
**Pull request by hadjri**
DEVPROD-31985

### Description
Reverts the original change from [DEVPROD-22723](https://jira.mongodb.org/browse/DEVPROD-22723) which set the task_create_iso time to the task's create time to prevent issues where tasks running across midnight would have incomplete test results in the UI.

Changed to instead mimic the original behavior, which is to set the timestamp to the moment that the test results are uploaded. However, this implementation should still prevent the [DEVPROD-22723](https://jira.mongodb.org/browse/DEVPROD-22723) bug by storing the time of the first test results upload, and using it for all subsequent uploads for the same task.

### Testing
Tested in staging manually during normal conditions, and verified the original bug of a task appending multiple test results before and after midnight in the same execution.

I created a task that writes some test results, sleeps for 10 minutes, and then writes some more test results. I ran the task at 11:51PM to ensure that results were written before and after the midnight line in staging. I verified after the task completed all test results were available as expected.

### CODE REVIEW

File: agent/command/results_utils.go [hunk 1]
```diff
@@ -93,7 +93,11 @@ func attachTestResults(ctx context.Context, conf *internal.TaskConfig, td client
 }
 
 func uploadTestResults(ctx context.Context, comm client.Communicator, conf *internal.TaskConfig, results []testresult.TestResult, td client.TaskData, output *task.TaskOutput) (bool, error) {
-	createdAt := conf.Task.CreateTime
+	createdAt := conf.TestResultsCreatedAt
```
Comment (malikchaya2) [inline agent/command/results_utils.go:96]: Can you please add more details on how you verified in staging that the original bug of a task appending multiple test results before and after midnight in the same execution is fixed?

**Review by malikchaya2 — CHANGES_REQUESTED** [review]

**Comment by hadjri** [comment]
This was not ready for review. Github auto requested for review, removed it for now.

File: agent/command/results_utils.go [hunk 2]
```diff
@@ -93,7 +93,11 @@ func attachTestResults(ctx context.Context, conf *internal.TaskConfig, td client
...
+	if createdAt.IsZero() {
+		createdAt = time.Now()
+		conf.TestResultsCreatedAt = createdAt
+	}
```
Comment (malikchaya2) [inline agent/command/results_utils.go:100]: If a task: 
- uploads it's first test result on 3/1 at 11:59, 
- the pipeline runs on 3/2 at 2:00am and uploads everything from 3/1
- the task continues running and uploads another test result on 3/2 at 2:01 am, using the TestResultsCreatedAt of 11:59 

will we hit a similar issue of the pipeline missing some data?

File: agent/command/results_utils.go (hunk 1 above)
Comment (hadjri) [reply agent/command/results_utils.go:99]: Yep, updated the details!

File: agent/command/results_utils.go (hunk 2 above)
Comment (hadjri) [reply agent/command/results_utils.go:103]: Yes, though that would be an edge case that requires a task that: (1) uploads some test results, (2) continues running for hours past the 2 AM pipeline run, and (3) uploads more results, which is a  narrow window relative to the original systemic 35% miss rate from CreateTime DPIPE reported.

If they so choose, they can mitigate that on their end by running the pipeline 1 day behind and effectively eliminate that edge case. Since all uploads for a task land in the same partition (the date of the first upload), if DPIPE ingests 1 day behind, then even a task that started uploading on 3/1 at 11:59 PM and kept appending results into 3/2 would be fully finished by 3/2 end of day.                                                                                                                                   
                                                                                                                                                                                                                 
Then, the only tasks this wouldn't cover are ones that upload test results over a span of more than 1 day, which shouldn't realistically happen.

Comment (malikchaya2) [reply agent/command/results_utils.go:103]: I initially offered a fix that would rely on extend the timeline but that wasn't an acceptable fix for the team. I don't think this will be a risk they are willing to accept but I'm happy to defer to them.

Comment (hadjri) [reply agent/command/results_utils.go:103]: IIRC that concern was a little different, which was that running a delayed pipeline when `task_create_iso` reflects task _create_ time is not reliable because you never know if someone is going to activate / restart an old task from weeks or even months ago. So even running a lag of like 30 days would still likely miss test results.

Whereas with this, running a lag of just 1 day would be enough to reasonably cover this edge case, which is already niche

Comment (malikchaya2) [reply agent/command/results_utils.go:103]: Update: Confirmed with the data team over slack that this solution doesn't work for them because they need 100% coverage of all tasks.

**Review by malikchaya2 — CHANGES_REQUESTED** [review]

**Comment by hadjri** [comment]
Keeping the current solution as is while adding an explicit disallow for appending test results >24h apart, see https://mongodb.slack.com/archives/C69UXN1CP/p1776897748544379?thread_ts=1776470006.148739&cid=C69UXN1CP

File: agent/command/results_utils.go [hunk 3]
```diff
@@ -92,8 +92,17 @@ func attachTestResults(ctx context.Context, conf *internal.TaskConfig, td client
...
+
 func uploadTestResults(ctx context.Context, comm client.Communicator, conf *internal.TaskConfig, results []testresult.TestResult, td client.TaskData, output *task.TaskOutput) (bool, error) {
-	createdAt := conf.Task.CreateTime
+	createdAt := conf.TestResultsCreatedAt
+	if createdAt.IsZero() {
+		createdAt = time.Now()
+		conf.TestResultsCreatedAt = createdAt
+	}
+	if time.Since(conf.TestResultsCreatedAt) > maxTestResultsInterval {
+		return false, errors.Errorf("cannot append test results more than %s after the first upload", maxTestResultsInterval)
```
Comment (malikchaya2) [inline agent/command/results_utils.go]: To avoid threading, replying to [this](https://github.com/evergreen-ci/evergreen/pull/10124#issuecomment-4300420565) here. 

I disagree with us purposely breaking something that would otherwise go forward in order to keep the data within our expectations. 

We should respect what our `Max Task Exec Timeout (secs)` is currently set to as well as the user expectation is that if they provide us with test results to append it gets appended. Our system is currently built to allow for a task to run for `Max Task Exec Timeout` and to upload test results at whatever interval it chooses to configure it for. We should honor that even if it is an edge case that may never happen, or we should have a larger conversation around changing those expectations. 

This doesn't currently represent an error state and I believe erroring for this case is not the right course of action.

**Review by malikchaya2 — CHANGES_REQUESTED** [review]

File: agent/command/results_utils.go (hunk 3 above)
Comment (hadjri) [reply agent/command/results_utils.go]: Sharing the relevant discussion from slack regarding this:

> The scenario (a task appending test results more than 24 hours after its first upload) would be a very bespoke scenario and has never happened as of today. If codifying an invariant that already exists in practice can give DPIPE a assurance / guarantee about test result partitions, then I think it's worth it. If some task does hit this in the future, the error message is explicit about why, and the workaround is just to batch all test result uploads at the end of the task rather than incrementally over multiple days. cc [@annie.black](https://mongodb.enterprise.slack.com/team/UCBGNB7JP) if you have thoughts here.
> 
> It is an established pattern in Evergreen that we prevent / cap / limit people from certain actions in order to keep the stability and integrity of our system.
> 
> 1 in 1.5 million tasks run for over 24h, and out of that fraction, none have sent test results across a period spanning greater than 24h. So this is an exceedingly rare scenario. If we prevent it, it wouldn't impact any existing UX. Furthermore, there isn't a real reason to append test results across such a large gap, a task could just add all results at the very end and it would have the same effect.
> 
> Agree with Kelly, there is nobody using or asking for this niche edge case at the present (and I am skeptical if there is a good reason anyone would ever need it). Meanwhile on the other hand, we do have a real ask from folks here. Explicitly preventing this would give peace of mind to DPIPE and ensure they can catch absolutely 100% of test results.
> 
> So I am in favor of keeping the PR the way it currently is (It gets us to 100% AND fixes the [DEVPROD-22723](https://jira.mongodb.org/browse/DEVPROD-22723) bug by disallowing the edge case). Should for whatever reason we need to add support for this case in the future, we can work on a solution that supports it while keeps DPIPE at 100% at that time.

TLDR - this will not have any impact on existing UX and there is a clear workaround for anyone who might run into this edge case in the future.  Meanwhile we do have a real ask from the data team to do this for their peace-of-mind.

Comment (malikchaya2) [reply agent/command/results_utils.go]: Assigning @ablack12 for review

Comment (ablack12) [reply agent/command/results_utils.go]: Removing myself -- I'm good with the solution as we've talked about it and don't think I need to review the code here if it aligns with the decision we agreed on, and a follow-up ticket will be made to remove the user restriction.

Comment (malikchaya2) [reply agent/command/results_utils.go]: I'd like to make one small suggestion because I think it will address both concerns cleanly . 

How would you feel about this: Instead of returning an error here, we can log it and add an alert. We don't expect to hit it but if we do, we can let the data team know so that they know to backfill any data needed on their end and we can reach out to the user and ask them to change their flow to upload all test results at the end.

Comment (hadjri) [reply agent/command/results_utils.go]: That sounds good, I added it as an alert so we will be notified in the future 👍 

I filed [DEVPROD-32331](https://jira.mongodb.org/browse/DEVPROD-32331) to cover the work required to enable this workflow.

So ultimately this PR does the following:
- Reverts the original change from [DEVPROD-22723](https://jira.mongodb.org/browse/DEVPROD-22723) which set the task_create_iso time to the task's create time to prevent issues where tasks running across midnight would have incomplete test results in the UI
- Disallows test results from being appended for over a 24-hour consecutive window in the same task execution
- Alerts if this scenario occurs so we can reach out to the user & DPIPE as you mentioned

Let me know if this works! And I know this thread has gotten long so if needed I'm happy to jump on a quick zoom. Thanks for hanging in on this review

Comment (malikchaya2) [inline agent/command/results_utils.go]:
```suggestion
		return false, errors.Errorf("Cannot append test results more than %s after the first upload. Consider uploading all test results at the end of the task.", maxTestResultsInterval)
```

**Review by malikchaya2 — APPROVED** [review]

File: agent/command/results_utils.go (hunk 3 above)
Comment (malikchaya2) [reply agent/command/results_utils.go]: Awesome thank you! 
I was initially suggesting using the alert instead of the error and letting the data team know to ingest the data for that specific task until the user fixes it, but this flow works as well and wouldn't require letting the data team know since the data would reflect the state of the task accurately. We'd only have to let the user know how to fix the broken task. 

This looks good, let's remember to document what to do for this alert in our playbook.
